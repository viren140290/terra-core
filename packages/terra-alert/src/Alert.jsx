import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Button from 'terra-button';
import IconAlert from 'terra-icon/lib/icon/IconAlert';
import IconError from 'terra-icon/lib/icon/IconError';
import IconWarning from 'terra-icon/lib/icon/IconWarning';
import IconRequired from 'terra-icon/lib/icon/IconRequired';
import IconDiamond from 'terra-icon/lib/icon/IconDiamond';
import IconInformation from 'terra-icon/lib/icon/IconInformation';
import IconSuccess from 'terra-icon/lib/icon/IconSuccess';
import './Alert.scss';

const AlertTypes = {
  ALERT: 'alert',
  ERROR: 'error',
  WARNING: 'warning',
  REQUIRED: 'required',
  ADVISORY: 'advisory',
  CUSTOM: 'custom',
  INFORMATION: 'information',
  CONFIRMATION: 'confirmation',
};

const propTypes = {
  /**
   * An action element to be added to the action section of the alert.
   */
  alertAction: PropTypes.element,
  /**
   * Child Nodes providing the message content for the alert. Can contain text and HTML.
   */
  children: PropTypes.PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
  /**
   * The icon to be used for an alert of type custom.
   */
  customIcon: PropTypes.element,
  /**
   * The status bar color to be used for an alert of type custom.
   */
  customStatusColor: PropTypes.string,
  /**
   * Callback function triggered when Dismiss button is clicked. The presence of this prop will cause the Dismiss button to be included on the alert.
   */
  onDismiss: PropTypes.func,
  /**
   * The title for the alert which will be bolded.
   */
  title: PropTypes.string,
  /**
   * The type of alert to be rendered. One of AlertTypes.ALERT, AlertTypes.ERROR, AlertTypes.WARNING, AlertTypes.REQUIRED, AlertTypes.ADVISORY,
   * AlertTypes.INFORMATION, AlertTypes.CONFIRMATION, AlertTypes.CUSTOM.
   * Import AlertTypes for access to these type strings.
   */
  type: PropTypes.oneOf([
    AlertTypes.ALERT,
    AlertTypes.ERROR,
    AlertTypes.WARNING,
    AlertTypes.REQUIRED,
    AlertTypes.ADVISORY,
    AlertTypes.CUSTOM,
    AlertTypes.INFORMATION,
    AlertTypes.CONFIRMATION,
  ]),
};

const defaultProps = {
  alertAction: null,
  children: '',
  customIcon: null,
  customStatusColor: '',
  onDismiss: null,
  title: '',
  type: AlertTypes.ALERT,
};

const getAlertClassName = (type) => {
  switch (type) {
    case AlertTypes.ALERT:
      return 'terra-Alert--alert';
    case AlertTypes.ERROR:
      return 'terra-Alert--error';
    case AlertTypes.WARNING:
      return 'terra-Alert--warning';
    case AlertTypes.REQUIRED:
      return 'terra-Alert--required';
    case AlertTypes.ADVISORY:
      return 'terra-Alert--advisory';
    case AlertTypes.INFORMATION:
      return 'terra-Alert--information';
    case AlertTypes.CONFIRMATION:
      return 'terra-Alert--confirmation';
    case AlertTypes.CUSTOM:
      return 'terra-Alert--custom';
    default:
      return '';
  }
};

const getAlertIcon = (type, customIcon) => {
  switch (type) {
    case AlertTypes.ALERT:
      return (<IconAlert height="1.3333rem" width="1.3333rem" />);
    case AlertTypes.ERROR:
      return (<IconError height="1.3333rem" width="1.3333rem" />);
    case AlertTypes.WARNING:
      return (<IconWarning height="1.3333rem" width="1.3333rem" />);
    case AlertTypes.REQUIRED:
      return (<IconRequired height="1.3333rem" width="1.3333rem" />);
    case AlertTypes.ADVISORY:
      return (<IconDiamond height="1.3333rem" width="1.3333rem" />);
    case AlertTypes.INFORMATION:
      return (<IconInformation height="1.3333rem" width="1.3333rem" />);
    case AlertTypes.CONFIRMATION:
      return (<IconSuccess height="1.3333rem" width="1.3333rem" />);
    case AlertTypes.CUSTOM:
      return customIcon;
    default:
      return null;
  }
};

const getDefaultTitle = (type) => {
  switch (type) {
    case AlertTypes.ALERT:
      return 'Alert!';
    case AlertTypes.ERROR:
      return 'Error!';
    case AlertTypes.WARNING:
      return 'Warning!';
    case AlertTypes.REQUIRED:
      return 'Required!';
    case AlertTypes.ADVISORY:
      return 'Advisory!';
    case AlertTypes.INFORMATION:
      return 'Information!';
    case AlertTypes.CONFIRMATION:
      return 'Confirmed!';
    default:
      return '';
  }
};

const Alert = ({ type, children, title, customIcon, customStatusColor, onDismiss, alertAction, ...customProps }) => {
  const attributes = Object.assign({}, customProps);
  const alertTypeClassName = getAlertClassName(type);
  const AlertClassNames = classNames([
    'terra-Alert',
    alertTypeClassName,
    attributes.className,
  ]);

  let actionsSection = '';
  let dismissButton = '';
  const outerDivStyle = {};
  let alertSectionClassName = 'terra-Alert-section';
  let alertActionsClassName = 'terra-Alert-actions';

  if (type === AlertTypes.CUSTOM) {
    // For custom alert, there is no color assigned to the box-shadow style since it is to be specified
    // in the customStatusColor prop.  The box-shadow style is defined in CSS in order to get the
    // bidirectionality via the mixin.  As per the W3C spec if the box-shadow does not have the color
    // defined, it will use the prevailing color style, so setting that here. But then we need to set
    // the color style for the alert content so that it doesn't pick up the custom status color.
    outerDivStyle.color = customStatusColor;
    alertSectionClassName += ' terra-Alert-section--custom';
    alertActionsClassName += ' terra-Alert-actions--custom';
  }

  if (onDismiss) {
    dismissButton = (<Button text="Dismiss" size="medium" variant="secondary" onClick={onDismiss} />);
  }
  if (onDismiss || alertAction) {
    actionsSection = (
      <div className={alertActionsClassName}>
        {alertAction || ''}
        {dismissButton}
      </div>
    );
  }

  return (
    <div {...attributes} className={AlertClassNames} style={outerDivStyle} >
      <div className="terra-Alert-body">
        <div className="terra-Alert-icon">{getAlertIcon(type, customIcon)}</div>
        <div className={alertSectionClassName}>
          <strong className="terra-Alert-title">{title || getDefaultTitle(type)}</strong>
          <div className="terra-Alert-content">
            {children}
          </div>
        </div>
      </div>
      {actionsSection}
    </div>
  );
};

Alert.propTypes = propTypes;
Alert.defaultProps = defaultProps;

export { AlertTypes };
export default Alert;
