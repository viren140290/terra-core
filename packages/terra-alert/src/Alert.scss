@import '~terra-mixins';
@import './variables';
@import './mixins';

.terra-Alert {
  align-items: stretch;
  display: flex;
  flex: 1;
  flex-direction: column;
  justify-content: space-between;
  overflow: hidden;
  width: 100%;

  @media screen and (min-width: 768px) {
    flex-direction: row;
  }
}

.terra-Alert-icon {
  flex-shrink: 0;
  @include terra-margin-end(0.5rem);
}

.terra-Alert-body {
  align-items: flex-start;
  display: flex;
  flex: 1 1 auto;
  flex-direction: row;
  margin: 1rem 0.6667rem;
  overflow: hidden;
  padding-top: 0.1rem;
  
  @media screen and (min-width: 768px) {
    @include terra-margin-end(0);
  }
}

.terra-Alert-section {
  flex: 1;
  padding-top: 0.2rem;
}

.terra-Alert-section--custom {
  @include terra-alert-color($terra-alert-custom-text-color);
}

.terra-Alert-title {
  @include terra-float-start();
  @include terra-margin-end(0.5833rem);
}

.terra-Alert-content {
  p {
    margin-top: 0;
  }

  p:last-child {
    margin-bottom: 0;
  }
}

.terra-Alert-actions {
  align-items: flex-start;
  display: flex;
  flex-direction: row;
  flex-shrink: 0;
  justify-content: flex-end;
  margin: 0.5833rem 0.5833rem 0.5833rem 0.8333rem;
}

.terra-Alert-actions--custom {
  @include terra-alert-color($terra-alert-custom-text-color);
}

.terra-Alert-actions > :first-child {
  @include terra-margin-end(0.8333rem);
}

.terra-Alert-actions > :last-child {
  @include terra-margin-end(0);
}


/* default skin */
.terra-Alert--default {
  background-color: #f4f4f4;
  box-shadow: 2px 0 0 #007cc3 inset;
}

.terra-Alert--alert {
  @include terra-alert-box-shadow($terra-alert-alert-shadow-color);
  @include terra-alert-background-color($terra-alert-alert-color-scheme);
}

.terra-Alert--error {
  @include terra-alert-box-shadow($terra-alert-error-shadow-color);
  @include terra-alert-background-color($terra-alert-error-color-scheme);
}

.terra-Alert--warning {
  @include terra-alert-box-shadow($terra-alert-warning-shadow-color);
  @include terra-alert-background-color($terra-alert-warning-color-scheme);
}

.terra-Alert--required {
  @include terra-alert-box-shadow($terra-alert-required-shadow-color);
  @include terra-alert-background-color($terra-alert-required-color-scheme);
}

.terra-Alert--advisory {
  @include terra-alert-box-shadow($terra-alert-advisory-shadow-color);
  @include terra-alert-background-color($terra-alert-advisory-color-scheme);
}

.terra-Alert--information {
  @include terra-alert-box-shadow($terra-alert-information-shadow-color);
  @include terra-alert-background-color($terra-alert-information-color-scheme);
}

.terra-Alert--confirmation {
  @include terra-alert-box-shadow($terra-alert-confirmation-shadow-color);
  @include terra-alert-background-color($terra-alert-confirmation-color-scheme);
}

.terra-Alert--custom {
  @include terra-alert-color($terra-alert-custom-shadow-color);
  @include terra-alert-custom-box-shadow();
  @include terra-alert-background-color($terra-alert-custom-color-scheme);
}

.terra-Alert--alert .terra-Alert-title {
  @include terra-alert-color($terra-alert-alert-title-color-scheme);
}

.terra-Alert--error .terra-Alert-title {
  @include terra-alert-color($terra-alert-error-title-color-scheme);
}

.terra-Alert--required .terra-Alert-icon {
  @include terra-alert-color($terra-alert-required-icon-color-scheme);
}
