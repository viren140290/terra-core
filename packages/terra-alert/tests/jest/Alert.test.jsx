import React from 'react';
import IconAlert from 'terra-icon/lib/icon/IconAlert';
import IconError from 'terra-icon/lib/icon/IconError';
import Alert, { AlertTypes } from '../../src/Alert';

describe('Alert with no props', () => {
  const defaultRender = <Alert />;

  // Snapshot Tests
  it('should render a default component', () => {
    const wrapper = shallow(defaultRender);
    expect(wrapper).toMatchSnapshot();
  });

  // Prop Tests
  it('should render an Alert of type alert when no props given including IconAlert icon and default alert title', () => {
    const wrapper = shallow(defaultRender);
    expect(wrapper.find(IconAlert)).toHaveLength(1);
    expect(wrapper.childAt(0).childAt(1).childAt(0).text()).toEqual('Alert!');
  });

  // Structure Tests
  it('should have the class terra-Alert and terra-Alert--alert', () => {
    const wrapper = shallow(defaultRender);
    expect(wrapper.prop('className')).toContain('terra-Alert');
    expect(wrapper.prop('className')).toContain('terra-Alert--alert');
  });

  it('default alert should have one child with className terr-Alert-body', () => {
    const wrapper = shallow(defaultRender);
    expect(wrapper.children()).toHaveLength(1);
    expect(wrapper.childAt(0).prop('className')).toContain('terra-Alert-body');
  });

  it('default alert body should have two children one for the icon section and one for the content section', () => {
    const wrapper = shallow(defaultRender);
    expect(wrapper.childAt(0).children()).toHaveLength(2);
    expect(wrapper.childAt(0).childAt(0).prop('className')).toContain('terra-Alert-icon');
    expect(wrapper.childAt(0).childAt(1).prop('className')).toContain('terra-Alert-section');
  });

  it('default alert content section should have two children one for the title and one for the content', () => {
    const wrapper = shallow(defaultRender);
    expect(wrapper.childAt(0).childAt(1).children()).toHaveLength(2);
    expect(wrapper.childAt(0).childAt(1).childAt(0).prop('className')).toContain('terra-Alert-title');
    expect(wrapper.childAt(0).childAt(1).childAt(0).type()).toEqual('strong');
    expect(wrapper.childAt(0).childAt(1).childAt(1).prop('className')).toContain('terra-Alert-content');
  });
});

describe('Alert of type alert with text content', () => {
  const alertText = 'This is an alert.';
  const basicAlertRender = <Alert type={AlertTypes.ALERT} >{alertText}</Alert>;

  // Snapshot Tests
  it('should render an Alert component of type alert', () => {
    const wrapper = shallow(basicAlertRender);
    expect(wrapper).toMatchSnapshot();
  });

  // Prop Tests
  it('should render an Alert of type alert including IconAlert icon and default alert title', () => {
    const wrapper = shallow(basicAlertRender);
    expect(wrapper.find(IconAlert)).toHaveLength(1);
    expect(wrapper.childAt(0).childAt(1).childAt(0).text()).toEqual('Alert!');
    expect(wrapper.childAt(0).childAt(1).childAt(1).text()).toEqual(alertText);
  });

  // Structure Tests
  it('should have the class terra-Alert and terra-Alert--alert', () => {
    const wrapper = shallow(basicAlertRender);
    expect(wrapper.prop('className')).toContain('terra-Alert');
    expect(wrapper.prop('className')).toContain('terra-Alert--alert');
  });
});

describe('Alert of type error with text content', () => {
  const alertText = 'This is an error.';
  const basicErrorRender = <Alert type={AlertTypes.ERROR} >{alertText}</Alert>;

  // Snapshot Tests
  it('should render an Alert component of type error', () => {
    const wrapper = shallow(basicErrorRender);
    expect(wrapper).toMatchSnapshot();
  });

  // Prop Tests
  it('should render an Alert of type error including IconError icon and default error title', () => {
    const wrapper = shallow(basicErrorRender);
    expect(wrapper.find(IconError)).toHaveLength(1);
    expect(wrapper.childAt(0).childAt(1).childAt(0).text()).toEqual('Error!');
    expect(wrapper.childAt(0).childAt(1).childAt(1).text()).toEqual(alertText);
  });

  // Structure Tests
  it('should have the class terra-Alert and terra-Alert--error', () => {
    const wrapper = shallow(basicErrorRender);
    expect(wrapper.prop('className')).toContain('terra-Alert');
    expect(wrapper.prop('className')).toContain('terra-Alert--error');
  });
});
