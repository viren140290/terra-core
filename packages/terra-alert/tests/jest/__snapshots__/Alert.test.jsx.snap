exports[`Alert of type alert with text content should render an Alert component of type alert 1`] = `
<div
  className="terra-Alert terra-Alert--alert"
  style={Object {}}>
  <div
    className="terra-Alert-body">
    <div
      className="terra-Alert-icon">
      <IconAlert
        height="1.3333rem"
        viewBox="0 0 48 48"
        width="1.3333rem"
        xmlns="http://www.w3.org/2000/svg" />
    </div>
    <div
      className="terra-Alert-section">
      <strong
        className="terra-Alert-title">
        Alert!
      </strong>
      <div
        className="terra-Alert-content">
        This is an alert.
      </div>
    </div>
  </div>
</div>
`;

exports[`Alert of type error with text content should render an Alert component of type error 1`] = `
<div
  className="terra-Alert terra-Alert--error"
  style={Object {}}>
  <div
    className="terra-Alert-body">
    <div
      className="terra-Alert-icon">
      <IconError
        height="1.3333rem"
        viewBox="0 0 48 48"
        width="1.3333rem"
        xmlns="http://www.w3.org/2000/svg" />
    </div>
    <div
      className="terra-Alert-section">
      <strong
        className="terra-Alert-title">
        Error!
      </strong>
      <div
        className="terra-Alert-content">
        This is an error.
      </div>
    </div>
  </div>
</div>
`;

exports[`Alert with no props should render a default component 1`] = `
<div
  className="terra-Alert terra-Alert--alert"
  style={Object {}}>
  <div
    className="terra-Alert-body">
    <div
      className="terra-Alert-icon">
      <IconAlert
        height="1.3333rem"
        viewBox="0 0 48 48"
        width="1.3333rem"
        xmlns="http://www.w3.org/2000/svg" />
    </div>
    <div
      className="terra-Alert-section">
      <strong
        className="terra-Alert-title">
        Alert!
      </strong>
      <div
        className="terra-Alert-content" />
    </div>
  </div>
</div>
`;
